presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewRewriteQueryShapeValidator.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+/**
+ * Validates that the {@link QuerySpecification} provided to
+ * {@link MaterializedViewRewriteQueryShapeValidator#validateMaterializedViewOptimizationQueryShape}
+ * is able to be rewritten from a backing materialized view. This does not guarantee the
+ * QuerySpecification has a compatible materialized view to be rewritten with, only that
+ * its shape is not prohibitive to the optimization.
+ */
+public class MaterializedViewRewriteQueryShapeValidator
+        extends DefaultTraversalVisitor<Node, Void>
+{
+    private static final QualifiedName MIN = QualifiedName.of("MIN");
+    private static final QualifiedName MAX = QualifiedName.of("MAX");
+    private static final QualifiedName SUM = QualifiedName.of("SUM");
+    private static final QualifiedName COUNT = QualifiedName.of("COUNT");
+    private static final Set<QualifiedName> SUPPORTED_FUNCTION_CALLS = ImmutableSet.of(MIN, MAX, SUM, COUNT);
+
+    private QueryShapeContext queryShapeContext = new QueryShapeContext(true, Optional.empty());
+
+    public QueryShapeContext validateMaterializedViewOptimizationQueryShape(QuerySpecification querySpecification)
+    {
+        process(querySpecification);
+        return queryShapeContext;
+    }
+
+    @Override
+    protected Node visitFunctionCall(FunctionCall node, Void context)
+    {
+        QualifiedName name = node.getName();
+        if (!SUPPORTED_FUNCTION_CALLS.contains(name)) {
+            queryShapeContext = new QueryShapeContext(
+                    false, Optional.of(format("Query shape invalid: %s function is not supported for materialized view optimizations", node.getName())));
+        }
+        return super.visitFunctionCall(node, context);
+    }
+
+    public static class QueryShapeContext
+    {
+        private boolean isValid;
+        private Optional<String> message;
```

Comment:
Call it errorMessage, and if its not present, it means its not valid, so you dont need to store a separate boolean